exist-core/src/main/java/org/exist/xquery/functions/fn/transform/URIResolution.java

```diff
@@ -0,0 +1,139 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package org.exist.xquery.functions.fn.transform;
+
+import org.exist.dom.persistent.NodeProxy;
+import org.exist.security.PermissionDeniedException;
+import org.exist.xmldb.XmldbURI;
+import org.exist.xquery.ErrorCodes;
+import org.exist.xquery.Expression;
+import org.exist.xquery.XPathException;
+import org.exist.xquery.XQueryContext;
+import org.exist.xquery.util.DocUtils;
+import org.exist.xquery.value.AnyURIValue;
+import org.exist.xquery.value.Sequence;
+import org.exist.xquery.value.Type;
+import org.w3c.dom.Node;
+
+import javax.xml.transform.Source;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.URIResolver;
+import javax.xml.transform.dom.DOMSource;
+import java.net.URI;
+import java.net.URISyntaxException;
+
+public class URIResolution {
+
+    /**
+     * URI resolution, the core should be the same as for fn:resolve-uri
+     * @param relative URI to resolve
+     * @param base to resolve against
+     * @return resolved URI
+     * @throws URISyntaxException if resolution is not possible
+     */
+    static AnyURIValue resolveURI(final AnyURIValue relative, final AnyURIValue base) throws URISyntaxException, XPathException {
+        var relativeURI = new URI(relative.getStringValue());
+        if (relativeURI.isAbsolute()) {
+            return relative;
+        }
+        var baseURI = new URI(base.getStringValue() );
+        if (!baseURI.isAbsolute()) {
+            return relative;
+        }
+        try {
+            var xBase = XmldbURI.xmldbUriFor(baseURI);
+            var resolved = xBase.getURI().resolve(relativeURI);
+            return new AnyURIValue(XmldbURI.XMLDB_URI_PREFIX + resolved);
+        } catch (URISyntaxException e) {
+            return new AnyURIValue(baseURI.resolve(relativeURI));
+        }
+    }
+
+    public static class CompileTimeURIResolver implements URIResolver {
+
+        private final XQueryContext xQueryContext;
+        private final Expression containingExpression;
+
+        public CompileTimeURIResolver(XQueryContext xQueryContext, Expression containingExpression) {
+            this.xQueryContext = xQueryContext;
+            this.containingExpression = containingExpression;
+        }
+
+        @Override
+        public Source resolve(final String href, final String base) throws TransformerException {
+
+            try {
+                final AnyURIValue baseURI = new AnyURIValue(base);
+                final AnyURIValue hrefURI = new AnyURIValue(href);
+                var resolved = resolveURI(hrefURI, baseURI);
+                return resolveDocument(resolved.getStringValue());
+            } catch (URISyntaxException e) {
+                throw new TransformerException("Failed to resolve " +
+                    href +
+                    " against " +
+                    base, e);
+            } catch (org.exist.xquery.XPathException e) {
```

Comment:
is this FQN needed here? (more locations)

Reply:
I'm not clear what you mean. I have tried to put as much information as possible in the potential exceptions. Is this too much, or otherwise unreasonable ?

Reply:
I think @dizzzz might be suggesting that you could change `org.exist.xquery.XPathException` to just `XPathException` and import it. I haven't checked this though!

Reply:
Haha, of course, my brain was thinking about fully qualified URIs. doh!